File: src/main/scala/perfio/BufferUtil.scala
package perfio

import java.lang.invoke.MethodHandles
import java.nio.ByteOrder
import scala.reflect.ClassTag

private object BufferUtil {
  /** Compute a new buffer size for the given size alignment (and assuming the current size respects
   * this alignment, clamped at the maximum aligned value <= Int.MaxValue */
  def growBuffer(current: Int, target: Int, align: Int): Int = {
    var l = current
    while(l < target) {
      l *= 2
      if(l < 0) return Int.MaxValue - align + 1
    }
    l
  }

  val BA_LONG_BIG = bavh[Long](true)
  val BA_INT_BIG = bavh[Int](true)
  val BA_SHORT_BIG = bavh[Short](true)
  val BA_CHAR_BIG = bavh[Char](true)
  val BA_DOUBLE_BIG = bavh[Double](true)
  val BA_FLOAT_BIG = bavh[Float](true)

  val BA_LONG_LITTLE = bavh[Long](false)
  val BA_INT_LITTLE = bavh[Int](false)
  val BA_SHORT_LITTLE = bavh[Short](false)
  val BA_CHAR_LITTLE = bavh[Char](false)
  val BA_DOUBLE_LITTLE = bavh[Double](false)
  val BA_FLOAT_LITTLE = bavh[Float](false)

  @inline private[this] def bavh[T](@inline be: Boolean)(implicit @inline ct: ClassTag[Array[T]]) =
    MethodHandles.byteArrayViewVarHandle(ct.runtimeClass, if(be) ByteOrder.BIG_ENDIAN else ByteOrder.LITTLE_ENDIAN)

  val BB_LONG_BIG = bbvh[Long](true)
  val BB_INT_BIG = bbvh[Int](true)
  val BB_SHORT_BIG = bbvh[Short](true)
  val BB_CHAR_BIG = bbvh[Char](true)
  val BB_DOUBLE_BIG = bbvh[Double](true)
  val BB_FLOAT_BIG = bbvh[Float](true)

  val BB_LONG_LITTLE = bbvh[Long](false)
  val BB_INT_LITTLE = bbvh[Int](false)
  val BB_SHORT_LITTLE = bbvh[Short](false)
  val BB_CHAR_LITTLE = bbvh[Char](false)
  val BB_DOUBLE_LITTLE = bbvh[Double](false)
  val BB_FLOAT_LITTLE = bbvh[Float](false)

  @inline private[this] def bbvh[T](@inline be: Boolean)(implicit @inline ct: ClassTag[Array[T]]) =
    MethodHandles.byteBufferViewVarHandle(ct.runtimeClass, if(be) ByteOrder.BIG_ENDIAN else ByteOrder.LITTLE_ENDIAN)
}

File: src/main/java/perfio/BufferUtil.java
package perfio;

import java.io.IOException;
import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.invoke.MethodHandles;
import java.lang.invoke.VarHandle;
import java.nio.ByteOrder;
import java.nio.channels.FileChannel;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;

class BufferUtil {
  private BufferUtil() {}

  /// Compute a new buffer size for the given size alignment (and assuming the current size respects
  /// this alignment, clamped at the maximum aligned value <= [Integer#MAX_VALUE].
  static int growBuffer(int current, int target, int align) {
    int l = current;
    while(l < target) {
      l *= 2;
      if(l < 0) return Integer.MAX_VALUE - align + 1;
    }
    return l;
  }

  static final VarHandle BA_LONG_BIG = bavh(Long.TYPE, true);
  static final VarHandle BA_INT_BIG = bavh(Integer.TYPE, true);
  static final VarHandle BA_SHORT_BIG = bavh(Short.TYPE, true);
  static final VarHandle BA_CHAR_BIG = bavh(Character.TYPE, true);
  static final VarHandle BA_DOUBLE_BIG = bavh(Double.TYPE, true);
  static final VarHandle BA_FLOAT_BIG = bavh(Float.TYPE, true);

  static final VarHandle BA_LONG_LITTLE = bavh(Long.TYPE, false);
  static final VarHandle BA_INT_LITTLE = bavh(Integer.TYPE, false);
  static final VarHandle BA_SHORT_LITTLE = bavh(Short.TYPE, false);
  static final VarHandle BA_CHAR_LITTLE = bavh(Character.TYPE, false);
  static final VarHandle BA_DOUBLE_LITTLE = bavh(Double.TYPE, false);
  static final VarHandle BA_FLOAT_LITTLE = bavh(Float.TYPE, false);

  private static VarHandle bavh(Class<?> cls, boolean be) {
    return MethodHandles.byteArrayViewVarHandle(cls.arrayType(), be ? ByteOrder.BIG_ENDIAN : ByteOrder.LITTLE_ENDIAN);
  }

  static final VarHandle BB_LONG_BIG = bbvh(Long.TYPE, true);
  static final VarHandle BB_INT_BIG = bbvh(Integer.TYPE, true);
  static final VarHandle BB_SHORT_BIG = bbvh(Short.TYPE, true);
  static final VarHandle BB_CHAR_BIG = bbvh(Character.TYPE, true);
  static final VarHandle BB_DOUBLE_BIG = bbvh(Double.TYPE, true);
  static final VarHandle BB_FLOAT_BIG = bbvh(Float.TYPE, true);

  static final VarHandle BB_LONG_LITTLE = bbvh(Long.TYPE, false);
  static final VarHandle BB_INT_LITTLE = bbvh(Integer.TYPE, false);
  static final VarHandle BB_SHORT_LITTLE = bbvh(Short.TYPE, false);
  static final VarHandle BB_CHAR_LITTLE = bbvh(Character.TYPE, false);
  static final VarHandle BB_DOUBLE_LITTLE = bbvh(Double.TYPE, false);
  static final VarHandle BB_FLOAT_LITTLE = bbvh(Float.TYPE, false);

  private static VarHandle bbvh(Class<?> cls, boolean be) {
    return MethodHandles.byteBufferViewVarHandle(cls.arrayType(), be ? ByteOrder.BIG_ENDIAN : ByteOrder.LITTLE_ENDIAN);
  }

  static MemorySegment mapReadOnlyFile(Path file) throws IOException {
    var a = Arena.ofAuto();
    try (var ch = FileChannel.open(file, StandardOpenOption.READ)) {
      return ch.map(FileChannel.MapMode.READ_ONLY, 0, ch.size(), a);
    }
  }

  static final boolean VECTOR_ENABLED;
  static final int VECTOR_LENGTH;
  static {
    var e = false;
    var vlen = 8; // minimum for computing a usable minimum buffer length
    try {
      // Just a sanity check. We accept any reasonable size. Even 64-bit vectors (SWAR) are faster than scalar.
      // Hopefully this will guarantee that the preferred species is actually vectorized (which is not the case
      // with the experimental preview API at the moment).
      e = VectorSupport.SPECIES.length() >= 8 && !"true".equals(System.getProperty("perfio.disableVectorized"));
      if(e) vlen = VectorSupport.SPECIES.length();
    } catch(NoClassDefFoundError t) {}
    VECTOR_ENABLED = e;
    VECTOR_LENGTH = vlen;
  }
}
